dwave/optimization/model.py

```diff
@@ -0,0 +1,499 @@
+
+__all__ = ["Model"]
+
+
+@contextlib.contextmanager
+def locked(model: _Graph):
+    """Context manager that hold a locked model and unlocks it when the context is exited."""
+    try:
+        yield
+    finally:
+        model.unlock()
+
+
+class Model(_Graph):
+    """Nonlinear model.
+
+    The nonlinear model represents a general optimization problem with an
+    :term:`objective function` and/or constraints over variables of various
+    types.
+
+    The :class:`.Model` class can contain this model and its methods provide
+    convenient utilities for working with representations of a problem.
+
+    Examples:
+        This example creates a model for a
+        :class:`flow-shop-scheduling <dwave.optimization.generators.flow_shop_scheduling>`
+        problem with two jobs on three machines.
+
+        >>> from dwave.optimization.generators import flow_shop_scheduling
+        ...
+        >>> processing_times = [[10, 5, 7], [20, 10, 15]]
+        >>> model = flow_shop_scheduling(processing_times=processing_times)
+    """
+
+    objective: typing.Optional[ArraySymbol]
+    """Objective to be minimized.
+
+    Examples:
+        This example prints the value of the objective of a model representing
+        the simple polynomial, :math:`y = i^2 - 4i`, for a state with value
+        :math:`i=2.0`.
+
+        >>> from dwave.optimization import Model
+        ...
+        >>> model = Model()
+        >>> i = model.integer(lower_bound=-5, upper_bound=5)
+        >>> c = model.constant(4)
+        >>> y = i**2 - c*i
+        >>> model.minimize(y)
+        >>> with model.lock():
+        ...     model.states.resize(1)
+        ...     i.set_state(0, 2.0)
+        ...     print(f"Objective = {model.objective.state(0)}")
+        Objective = -4.0
+    """
+
+    states: States
+    """States of the model.
+
+    :ref:`States <intro_optimization_states>` represent assignments of values
+    to a symbol.
+
+    See also:
+        :ref:`States methods <optimization_models>` such as
+        :meth:`~dwave.optimization.model.States.size` and
+        :meth:`~dwave.optimization.model.States.resize`.
+    """
+
+    def __init__(self):
+        self.objective = None
+        self.states = States(self)
+
+    def binary(self, shape: typing.Optional[_ShapeLike] = None) -> BinaryVariable:
+        r"""Create a binary symbol as a decision variable.
+
+        Args:
+            shape: Shape of the binary array to create.
+
+        Returns:
+            A binary symbol.
+
+        Examples:
+            This example creates a :math:`1 \times 20`-sized binary symbol.
+
+            >>> from dwave.optimization.model import Model
+            >>> model = Model()
+            >>> x = model.binary((1,20))
+        """
+        from dwave.optimization.symbols import BinaryVariable  # avoid circular import
+        return BinaryVariable(self, shape)
+
+    def constant(self, array_like: numpy.typing.ArrayLike) -> Constant:
```

Comment:
This and maybe `to_networkx` are the only methods I believe that we'll want to have on the future `Expression` class that we would duplicate. Considering how short this is and that `to_networkx` is not necessary for now (if ever) seems fine to me.